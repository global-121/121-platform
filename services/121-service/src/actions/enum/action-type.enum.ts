export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
