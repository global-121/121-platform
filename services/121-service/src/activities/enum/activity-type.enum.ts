export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
