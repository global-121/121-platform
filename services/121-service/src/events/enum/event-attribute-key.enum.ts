export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateWithRegistrationId = 'duplicateWithRegistrationId',
  duplicateWithRegistrationProgramId = 'duplicateWithRegistrationProgramId',
}
