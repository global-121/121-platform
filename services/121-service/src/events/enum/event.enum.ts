export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplicate = 'ignoredDuplicate',
}
