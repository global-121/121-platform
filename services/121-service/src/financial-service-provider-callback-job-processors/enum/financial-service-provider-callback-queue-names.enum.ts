export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
