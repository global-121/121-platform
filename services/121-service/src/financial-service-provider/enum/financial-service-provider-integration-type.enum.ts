export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
