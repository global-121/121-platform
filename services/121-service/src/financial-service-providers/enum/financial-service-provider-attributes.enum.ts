export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
