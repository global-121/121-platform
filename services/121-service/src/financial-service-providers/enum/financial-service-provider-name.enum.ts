export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
