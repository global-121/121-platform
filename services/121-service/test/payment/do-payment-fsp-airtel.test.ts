import { HttpStatus } from '@nestjs/common';

import { TransactionStatusEnum } from '@121-service/src/payments/transactions/enums/transaction-status.enum';
import { SeedScript } from '@121-service/src/scripts/enum/seed-script.enum';
import {
  doPayment,
  getTransactions,
  waitForPaymentTransactionsToComplete,
} from '@121-service/test/helpers/program.helper';
import { seedIncludedRegistrations } from '@121-service/test/helpers/registration.helper';
import {
  getAccessToken,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import { registrationAirtel } from '@121-service/test/registrations/pagination/pagination-data';

const programId = 1;
const amount = 200;

describe('Do payment with FSP: Airtel', () => {
  let accessToken: string;
  beforeAll(async () => {
    await resetDB(SeedScript.airtelProgram, __filename);
    accessToken = await getAccessToken();
  });

  it('should succesfully initiate a payment', async () => {
    // Arrange
    const registrationAirtelInitiatePayment = {
      ...registrationAirtel,
      referenceId: 'airtel-initiate-payment',
    };
    const paymentReferenceIds = [registrationAirtelInitiatePayment.referenceId];

    await seedIncludedRegistrations(
      [registrationAirtelInitiatePayment],
      programId,
      accessToken,
    );

    // Act
    const doPaymentResponse = await doPayment({
      programId,
      amount,
      referenceIds: paymentReferenceIds,
      accessToken,
    });
    const paymentId = doPaymentResponse.body.id;

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      paymentId,
      accessToken,
      maxWaitTimeMs: 30_000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
        TransactionStatusEnum.waiting,
      ],
    });

    // Assert
    expect(doPaymentResponse.status).toBe(HttpStatus.ACCEPTED);
    expect(doPaymentResponse.body.applicableCount).toBe(
      paymentReferenceIds.length,
    );
    expect(doPaymentResponse.body.totalFilterCount).toBe(
      paymentReferenceIds.length,
    );
    expect(doPaymentResponse.body.nonApplicableCount).toBe(0);
    expect(doPaymentResponse.body.sumPaymentAmountMultiplier).toBe(
      registrationAirtelInitiatePayment.paymentAmountMultiplier,
    );
  });
  // Arrange

  it('should successfully do a payment', async () => {
    // Arrange
    const registrationAirtelSuccessTransaction = {
      ...registrationAirtel,
      referenceId: 'airtel-success-transaction',
    };

    const paymentReferenceIds = [
      registrationAirtelSuccessTransaction.referenceId,
    ];
    await seedIncludedRegistrations(
      [registrationAirtelSuccessTransaction],
      programId,
      accessToken,
    );

    // Act
    const doPaymentResponse = await doPayment({
      programId,
      amount,
      referenceIds: paymentReferenceIds,
      accessToken,
    });
    const paymentId = doPaymentResponse.body.id;

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      paymentId,
      accessToken,
      maxWaitTimeMs: 30_000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
        TransactionStatusEnum.waiting,
      ],
    });
    const getTransactionsResult = await getTransactions({
      programId,
      paymentId,
      registrationReferenceId: registrationAirtelSuccessTransaction.referenceId,
      accessToken,
    });
    const transaction = getTransactionsResult.body[0];

    // Assert
    expect(transaction.errorMessage).toBe(null);
    expect(transaction.status).toBe(TransactionStatusEnum.success);
  });

  it('should correctly handle duplicate Airtel transactionId', async () => {
    // Arrange
    const registrationAirtelDuplicateTransactionId = {
      ...registrationAirtel,
      phoneNumber: '260000000001',
      referenceId: 'airtel-duplicate-transaction-id',
    };
    const paymentReferenceIds = [
      registrationAirtelDuplicateTransactionId.referenceId,
    ];
    await seedIncludedRegistrations(
      [registrationAirtelDuplicateTransactionId],
      programId,
      accessToken,
    );

    // Act
    const doPaymentResponse = await doPayment({
      programId,
      amount,
      referenceIds: paymentReferenceIds,
      accessToken,
    });
    const paymentId = doPaymentResponse.body.id;

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      paymentId,
      accessToken,
      maxWaitTimeMs: 30_000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
        TransactionStatusEnum.waiting,
      ],
    });
    const getTransactionsResult = await getTransactions({
      programId,
      paymentId,
      registrationReferenceId:
        registrationAirtelDuplicateTransactionId.referenceId,
      accessToken,
    });
    const transaction = getTransactionsResult.body[0];

    // Assert

    expect(transaction.errorMessage).toBe(null);
    expect(transaction.status).toBe(TransactionStatusEnum.success);
  });

  it('should create a transaction with status error when we do a payment with an invalid phonenumber', async () => {
    // Arrange
    const registrationAirtelDuplicateTransactionId = {
      ...registrationAirtel,
      phoneNumber: '260000000002',
      referenceId: 'airtel-invalid-phonenumber',
    };
    const paymentReferenceIds = [
      registrationAirtelDuplicateTransactionId.referenceId,
    ];
    await seedIncludedRegistrations(
      [registrationAirtelDuplicateTransactionId],
      programId,
      accessToken,
    );

    // Act
    const doPaymentResponse = await doPayment({
      programId,
      amount,
      referenceIds: paymentReferenceIds,
      accessToken,
    });
    const paymentId = doPaymentResponse.body.id;

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      paymentId,
      accessToken,
      maxWaitTimeMs: 30_000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
        TransactionStatusEnum.waiting,
      ],
    });
    const getTransactionsResult = await getTransactions({
      programId,
      paymentId,
      registrationReferenceId:
        registrationAirtelDuplicateTransactionId.referenceId,
      accessToken,
    });
    const transaction = getTransactionsResult.body[0];

    // Assert
    expect(transaction.errorMessage).toMatchSnapshot();
    expect(transaction.status).toBe(TransactionStatusEnum.error);
  });

  it('should create a transaction with status waiting when we get an Ambiguous error from Airtel', async () => {
    // Arrange
    const registrationAirtelAmbiguousError = {
      ...registrationAirtel,
      phoneNumber: '260000000003',
      referenceId: 'airtel-ambiguous-error',
    };
    const paymentReferenceIds = [registrationAirtelAmbiguousError.referenceId];
    await seedIncludedRegistrations(
      [registrationAirtelAmbiguousError],
      programId,
      accessToken,
    );

    // Act
    const doPaymentResponse = await doPayment({
      programId,
      amount,
      referenceIds: paymentReferenceIds,
      accessToken,
    });
    const paymentId = doPaymentResponse.body.id;

    await waitForPaymentTransactionsToComplete({
      programId,
      paymentReferenceIds,
      paymentId,
      accessToken,
      maxWaitTimeMs: 30_000,
      completeStatusses: [
        TransactionStatusEnum.success,
        TransactionStatusEnum.error,
        TransactionStatusEnum.waiting,
      ],
    });
    const getTransactionsResult = await getTransactions({
      programId,
      paymentId,
      registrationReferenceId: registrationAirtelAmbiguousError.referenceId,
      accessToken,
    });
    const transaction = getTransactionsResult.body[0];

    // Assert
    expect(transaction.status).toBe(TransactionStatusEnum.waiting);
    expect(transaction.errorMessage).toMatch(
      /^Airtel Error: Please use the Airtel Mobiquity portal to find out the status of the transaction. Airtel transaction id: [a-z0-9]{64} - Status: Ambiguous/,
    );
  });
});
