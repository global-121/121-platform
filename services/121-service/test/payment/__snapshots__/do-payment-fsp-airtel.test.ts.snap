// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Do payment with FSP: Airtel should create a transaction with status error when we do a payment with an invalid phonenumber 1`] = `"Airtel Error: Mobile number entered is incorrect. (DP00900001012)"`;
